packages/core/src/StateNode.ts

```diff
@@ -1268,19 +1268,31 @@ class StateNode<
       delete history.history;
     }
 
-    if (!resolvedStateValue) {
+    // There are transient transitions if the machine is not in a final state
+    // and if some of the state nodes have transient ("always") transitions.
+    const isTransient =
+      !isDone &&
+      (this._transient ||
+        configuration.some((stateNode) => {
+          return stateNode._transient;
+        }));
+
+    // If there are no enabled transitions, check if there are transient transitions.
+    // If there are transient transitions, continue checking for more transitions
+    // because an transient transition should be triggered even if there are no
+    // enabled transitions.
+    //
+    // If we're already working on an transient transition (by checking
+    // if the event is a NULL_EVENT), then stop to prevent an infinite loop.
```

Comment:
Actually... we should continue checking transient transitions **if** we have successfully selected a transition. Not sure how hard it would be to cover this in the current implementation though.

Reply:
What I meant by this is that the way we "trigger" transient transitions is by raising a `NULL_EVENT`, and if we didn't have this check here, we would have transient transitions infinitely since there would always be a state with a transient transition (if we didn't transition to a new state).

So this check is necessary. Try removing it ➡️  infinite loop 😱

Reply:
I've added a test for the case I had in mind - it works after all 🎉 I got confused by the fact that you didn't check `willTransition` here but as it turns out `resolvedStateValue` check was synonymous to it in this context.